logserver/src/main/java/alluxio/logserver/AlluxioLogServerProcess.java

```diff
@@ -0,0 +1,228 @@
+ * {@link AlluxioLogServerProcess} allows masters and workers to "push" their logs to a
+ * centralized log server where another copy of the logs will be stored.
+ */
+public class AlluxioLogServerProcess implements LogServerProcess {
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioLogServer.class);
+  private static final long STOP_TIMEOUT_IN_MS = 60000;
+  private static final int BASE_SLEEP_TIME_IN_MS = 50;
+  private static final int MAX_SLEEP_TIME_IN_MS = 30000;
+  private static final int MAX_NUM_RETRY = 20;
+
+  private final String mBaseLogsDir;
+  private int mPort;
+  private ServerSocket mServerSocket;
+  private final int mMinNumberOfThreads;
+  private final int mMaxNumberOfThreads;
+  private ExecutorService mThreadPool;
+  private volatile boolean mStopped;
+
+  /**
+   * Construct an {@link AlluxioLogServerProcess} instance.
+   *
+   * @param baseLogsDir base directory to store the logs pushed from remote Alluxio servers
+   */
+  public AlluxioLogServerProcess(String baseLogsDir) {
+    mPort = Configuration.getInt(PropertyKey.LOG_SERVER_PORT);
+    mMinNumberOfThreads = Configuration.getInt(PropertyKey.MASTER_WORKER_THREADS_MIN) + 2;
+    mMaxNumberOfThreads = Configuration.getInt(PropertyKey.MASTER_WORKER_THREADS_MAX) + 2;
+    mBaseLogsDir = baseLogsDir;
+    mStopped = true;
+  }
+
+  @Override
+  public void start() throws Exception {
+    startServing();
+  }
+
+  @Override
+  public void stop() throws Exception {
+    stopServing();
+  }
+
+  @Override
+  public void waitForReady() {
+    CommonUtils.waitFor(this + " to start", new Function<Void, Boolean>() {
+      @Override
+      public Boolean apply(Void input) {
+        return mStopped == false;
+      }
+    }, WaitForOptions.defaults().setTimeoutMs(10000));
+  }
+
+  /**
+   * Create and start logging server and client thread pool.
+   */
+  private void startServing() {
+    SynchronousQueue<Runnable> synchronousQueue =
+        new SynchronousQueue<>();
+    mThreadPool =
+        new ThreadPoolExecutor(mMinNumberOfThreads, mMaxNumberOfThreads,
+            STOP_TIMEOUT_IN_MS, TimeUnit.MILLISECONDS, synchronousQueue);
+    try {
+      mServerSocket = new ServerSocket(mPort);
+    } catch (IOException e) {
+      LOG.error("Failed to bind to port {}.", mPort);
+      return;
+    }
+    mStopped = false;
+    while (!mStopped) {
+      try {
+        Socket client = mServerSocket.accept();
+        InetAddress inetAddress = client.getInetAddress();
+        AlluxioLog4jSocketNode clientSocketNode =
+            new AlluxioLog4jSocketNode(this, client);
+        RetryPolicy retryPolicy = new ExponentialBackoffRetry(
+            BASE_SLEEP_TIME_IN_MS, MAX_SLEEP_TIME_IN_MS, MAX_NUM_RETRY);
+        while (true) {
+          try {
+            mThreadPool.execute(clientSocketNode);
+            break;
+          } catch (Throwable t) {
+            if (t instanceof RejectedExecutionException) {
+              if (!retryPolicy.attemptRetry()) {
```

Comment:
when would we hit rejected execution exception? I'd expect that it would only happen when there are too many clients, and retrying isn't going to solve that problem. It might make more sense to fail with an error message saying that the number of threads needs to be increased

Reply:
Done.